src/IROperator.cpp

```diff
@@ -480,6 +480,47 @@ void match_types(Expr &a, Expr &b) {
     }
 }
 
+void match_types_bitwise(Expr &x, Expr &y, const char *op_name) {
+    user_assert(x.defined() && y.defined()) << op_name << " of undefined Expr\n";
+    user_assert(x.type().is_int() || x.type().is_uint())
+      << "The first argument to " << op_name << " must be an integer or unsigned integer";
+    user_assert(y.type().is_int() || y.type().is_uint())
+      << "The second argument to " << op_name << " must be an integer or unsigned integer";
+
+    // Give explicit type to immediate constants in expression.
```

Comment:
We used to do this with match_types, and it caused problems. Consider:
```
Expr f(Expr a, Expr b) {
  return a & b;
}
```

Now you have a helper function which may error or change behavior in other ways, depending on the *representation* of b but not the value or (Halide or C++) type. E.g. if b is '2+3' then that will produce different behavior to if b is '5'

We went with overloads that take explicit ints for the other operators, so that the difference was determined by C++ type. We haven't had cause to regret that, so I think we should do the same here.

Reply:
Will implement.